### PR TITLE
PT-201: Add override for FromChangedEntry to accept operationType

### PR DESCRIPTION
**Pull request by eMazeika**
https://virtocommerce.atlassian.net/browse/PT-201

The method override can be called in order to specify the operationType.

### CODE REVIEW

**Comment by mvktsk** [comment]
Review task https://virtocommerce.atlassian.net/browse/VP-7280 has been created

File: src/VirtoCommerce.Platform.Core/ChangeLog/OperationLog.cs [hunk 1]
```diff
@@ -36,5 +37,21 @@ public virtual object Clone()
...
+        protected virtual string GetTypeNameForLogging<T>(T entry) where T : IEntity
+        {
+            if (entry is AuditableEntity)
+            {
+                return entry.GetType().GetTypeInheritanceChainTo(typeof(AuditableEntity)).Last().Name;
+            }
+            else
+            {
+                return entry.GetType().Name;
+            }
```
Comment (avas) [inline src/VirtoCommerce.Platform.Core/ChangeLog/OperationLog.cs:54]:
```suggestion
            else if (entry is Entity)
            {
                return entry.GetType().GetTypeInheritanceChainTo(typeof(Entity)).Last().Name;
            }
            else
            {
                return entry.GetType().Name;
            }
```
There are many types that do not derive the `AuditableEntity` but still can be extended - this approach will work consistently with these types.

File: src/VirtoCommerce.Platform.Core/ChangeLog/OperationLog.cs [hunk 2]
```diff
@@ -36,5 +37,21 @@ public virtual object Clone()
...
+
+        /// <summary>
+        /// Find the type name for operation log. It's a direct successor of AuditableEntity by default.
+        /// </summary>
+        /// <returns>Name of found type</returns>
+        protected virtual string GetTypeNameForLogging<T>(T entry) where T : IEntity
+        {
+            if (entry is AuditableEntity)
+            {
+                return entry.GetType().GetTypeInheritanceChainTo(typeof(AuditableEntity)).Last().Name;
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Core/ChangeLog/OperationLog.cs:49]: I also would like to check all modules that  call `IChangeLogSearchService.ChangeLogSearchCriteria` with the setting `ObjectType` parameter, in order to ensure that we don't introduce any breaking changes

File: src/VirtoCommerce.Platform.Core/ChangeLog/OperationLog.cs (hunk 1 above)
Comment (eMazeika) [reply src/VirtoCommerce.Platform.Core/ChangeLog/OperationLog.cs:58]: Interesting, that I could not find any descendant of Entity, that would be logged (All were AuditableEntity)

File: src/VirtoCommerce.Platform.Core/ChangeLog/OperationLog.cs (hunk 2 above)
Comment (tatarincev) [reply src/VirtoCommerce.Platform.Core/ChangeLog/OperationLog.cs:49]: In addition, I strongly recommend you to fix this issue on the customer module local level https://github.com/VirtoCommerce/vc-module-customer/blob/dev/src/VirtoCommerce.CustomerModule.Data/Handlers/LogChangesEventHandler.cs#L30

and do not touch platform code

Comment (eMazeika) [reply src/VirtoCommerce.Platform.Core/ChangeLog/OperationLog.cs:49]: In most cases, the modules will benefit from this change (all classes derive from AuditableEntity). One exception is CustomerOrder, which uses its own logic to create OperationLog (not using .FromChangedEntry<T>).

**Custom logic in vc-module-order and vc-module-avatax** modules:
https://github.com/VirtoCommerce/vc-module-order/blob/d6b5805e1b0adb955c5fdb768879a3dfa8e41f6f/src/VirtoCommerce.OrdersModule.Web/Controllers/Api/OrderModuleController.cs
ObjectTypes = allHasChangesObjects.Select(x => x.GetType().Name).Distinct().ToArray()
...
abstract class OrderOperation : AuditableEntity

https://github.com/VirtoCommerce/vc-module-avatax/blob/0283b2fd6d453561da7d2b37cac141fcfd1fe618/src/AvaTax.TaxModule.Data/Services/ChangeLogBasedOrdersFeed.cs#L115
var concreteOrder = AbstractTypeFactory<CustomerOrder>.TryCreateInstance();
return concreteOrder.GetType().Name;

**Other modules, where the change is compatible**:

https://github.com/VirtoCommerce/vc-module-customer/blob/dfa9cebace928ff9a0e87c036f2e3b25d9382ded/src/VirtoCommerce.CustomerModule.Data/Search/Indexing/MemberDocumentChangesProvider.cs#L16
Member : AuditableEntity
 
https://github.com/VirtoCommerce/vc-module-catalog-publishing/blob/6a6fab57e6e39135846f6add084a280072044e88/src/VirtoCommerce.CatalogPublishingModule.Data/Search/Indexing/ProductCompletenessChangesProvider.cs#L18
CompletenessEntryEntity : AuditableEntity

https://github.com/VirtoCommerce/vc-module-demo-features/blob/6a937d1a4f7b5362906aa5749a6e22e2a52579a6/src/VirtoCommerce.DemoSolutionFeaturesModule.Data/Search/Customer/DemoTaggedMemberIndexChangesProvider.cs#L14
DemoTaggedMember : AuditableEntity

https://github.com/VirtoCommerce/vc-module-inventory/blob/f2259c30f06e6c2f6c7f80a00aeeb744ab8ff4a1/src/VirtoCommerce.InventoryModule.Data/Search/Indexing/ProductAvailabilityChangesProvider.cs#L19
InventoryInfo : AuditableEntity

https://github.com/VirtoCommerce/vc-module-catalog/blob/1ecf74729589593b2095fec2b1a35bc1684f7be8/src/VirtoCommerce.CatalogModule.Data/Search/Indexing/CategoryDocumentChangesProvider.cs#L17
Category : AuditableEntity

https://github.com/VirtoCommerce/vc-module-catalog/blob/1ecf74729589593b2095fec2b1a35bc1684f7be8/src/VirtoCommerce.CatalogModule.Data/Search/Indexing/ProductDocumentChangesProvider.cs#L17
CatalogProduct : AuditableEntity

https://github.com/VirtoCommerce/vc-module-demo-customer-segments/blob/25d3006d88837b6dc8b0461a83085c5548286415/src/VirtoCommerce.DemoCustomerSegmentsModule.Data/Search/Indexing/DemoMemberDocumentChangesProvider.cs#L85
DemoCustomerSegment : AuditableEntity

https://github.com/VirtoCommerce/vc-module-catalog-personalization/blob/b6f841cf01cba04bcb97d22a880d179dfd829649/src/VirtoCommerce.CatalogPersonalizationModule.Data/Search/Indexing/TaggedItemIndexChangesProvider.cs#L16
TaggedItemEntity : AuditableEntity

https://github.com/VirtoCommerce/vc-module-pricing/blob/07c9104c11f12b3688cf29dfbd41e7ec3ff83e72/src/VirtoCommerce.PricingModule.Data/Search/ProductPriceDocumentChangesProvider.cs#L20
Price : AuditableEntity

**Review by Vectorfield4 — DISMISSED** [review]
Check for breaking changes, please.

**Review by Vectorfield4 — APPROVED** [review]

**Pull request merged**